src/dataprotection/azext_dataprotection/manual/custom.py

```diff
@@ -255,6 +255,132 @@ def dataprotection_backup_instance_list_from_resourcegraph(client, datasource_ty
     return response.data
 
 
+def dataprotection_backup_instance_update_msi_permissions(cmd, client, resource_group_name, datasource_type, vault_name, operation, permissions_scope, backup_instance=None, keyvault_id=None):
+    backup_vault = client.get(resource_group_name=resource_group_name,
+                              vault_name=vault_name)
+    principal_id = backup_vault.identity.principal_id
+
+    role_assignments_arr = []
+
+    if operation == 'Backup' and backup_instance is None:
+        raise CLIError("--backup-instance needs to be given when --operation is given as Backup")
+
+    if not keyvault_id and datasource_type == 'AzureDatabaseForPostgreSQL':
+        raise CLIError("--keyvault-id needs to be given when --datasource-type is AzureDatabaseForPostgreSQL")
+
+    from azure.cli.command_modules.role.custom import list_role_assignments, create_role_assignment
+
+    manifest = helper.load_manifest(datasource_type)
+    resource_map = {}
+
+    keyvault_client = None
+    keyvault = None
+    keyvault_subscription = None
+    keyvault_scope_arr = None
+    keyvault_name = None
+    keyvault_rg = None
+    if manifest['supportSecretStoreAuthentication']:
+        cmd.command_kwargs['operation_group'] = 'vaults'
+
+        from azure.cli.core.profiles import ResourceType
+        from azure.cli.command_modules.keyvault._client_factory import Clients
+        from azure.cli.core.commands.client_factory import get_mgmt_service_client
+
+        keyvault_scope_arr = keyvault_id.split("/")
```

Comment:
try making use of helper functions here. In helpers.py create functions like -

get_subscription_from_arm_id, get_rg_from_arm_id, get_name_from_arm_id etc. and use them as these functions are general and will get used at multiple places.